src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/EditorConfigNamingStyleParser.SymbolSpec.cs

```diff
@@ -0,0 +1,253 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles.SymbolSpecification;
+
+namespace Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles
+{
+    internal static partial class EditorConfigNamingStyleParser
+    {
+        private static bool TryGetSymbolSpec(
+            string namingRuleTitle,
+            IReadOnlyDictionary<string, object> allRawConventions,
+            out SymbolSpecification symbolSpec)
+        {
+            symbolSpec = null;
+            string symbolSpecName;
+            if (!TryGetSymbolSpecNameForNamingRule(namingRuleTitle, allRawConventions, out symbolSpecName))
+            {
+                return false;
+            }
+
+            var applicableKinds = GetSymbolsApplicableKinds(symbolSpecName, allRawConventions);
+            var applicableAccessibilities = GetSymbolsApplicableAccessibilities(symbolSpecName, allRawConventions);
+            var requiredModifiers = GetSymbolsRequiredModifiers(symbolSpecName, allRawConventions);
+
+            symbolSpec = new SymbolSpecification(
+                    Guid.NewGuid(),
+                    symbolSpecName,
+                    symbolKindList: applicableKinds,
+                    accessibilityKindList: applicableAccessibilities,
+                    modifiers: requiredModifiers);
+            return true;
+        }
+
+        private static bool TryGetSymbolSpecNameForNamingRule(
+            string namingRuleName,
+            IReadOnlyDictionary<string, object> allRawConventions,
+            out string symbolSpecName)
+        {
+            symbolSpecName = null;
+            var result = from kvp in allRawConventions
+                         where kvp.Key.Trim() == $"dotnet_naming_rule.{namingRuleName}.symbols"
+                         select kvp.Value;
+
+            if (result.IsSingle())
+            {
+                symbolSpecName = (string)result.Single();
+                return true;
+            }
+
+            return false;
+        }
+
+        private static IList<SymbolKindOrTypeKind> GetSymbolsApplicableKinds(
+            string symbolSpecName,
+            IReadOnlyDictionary<string, object> allRawConventions)
+        {
+            var result = from kvp in allRawConventions
+                         where kvp.Key.Trim() == $"dotnet_naming_symbols.{symbolSpecName}.applicable_kinds"
+                         select kvp.Value;
+
+            if (result.IsSingle())
+            {
+                return ParseSymbolKindList((string)result.Single()).ToList();
+            }
+
+            return SpecializedCollections.EmptyList<SymbolKindOrTypeKind>();
+        }
+
+        private static IEnumerable<SymbolKindOrTypeKind> ParseSymbolKindList(string symbolSpecApplicableKinds)
+        {
+            if (symbolSpecApplicableKinds == null)
+            {
+                yield break;
```

Comment:
How often will we be parsing these things?  I'm concerned about all the yield enumerators here for perf.  Can we switch over to `ImmutableArrayBuilder` for these instead?